### PR TITLE
FINERACT-1943 Add SubmittedOnDate to Savings Transaction

### PR DESCRIPTION
**Pull request by Mk9894**
## Description

Describe the changes made and why they were made.

Ignore if these details are present on the associated [Apache Fineract JIRA ticket](https://github.com/apache/fineract/pull/1284).

## Checklist

Please make sure these boxes are checked before submitting your pull request - thanks!

- [ ] Write the commit message as per https://github.com/apache/fineract/#pull-requests

- [ ] Acknowledge that we will not review PRs that are not passing the build _("green")_ - it is your responsibility to get a proposed PR to pass the build, not primarily the project's maintainers.

- [ ] Create/update unit or integration tests for verifying the changes made.

- [ ] Follow coding conventions at https://cwiki.apache.org/confluence/display/FINERACT/Coding+Conventions.

- [ ] Add required Swagger annotation and update API documentation at fineract-provider/src/main/resources/static/legacy-docs/apiLive.htm with details of any API changes

- [ ] Submission is not a "code dump".  (Large changes can be made "in repository" via a branch.  Ask on the developer mailing list for guidance, if required.)

FYI our guidelines for code reviews are at https://cwiki.apache.org/confluence/display/FINERACT/Code+Review+Guide.

### CODE REVIEW

**Review by marta-jankovics — APPROVED** [review]

**Comment by adamsaghy** [comment]
@Mk9894 Some tests are failing, kindly check them
```  Test testRunningBalanceAfterDepositReversalWithBackdateConfigurationOn() FAILED (3s)

  org.opentest4j.AssertionFailedError: Equality check for Balance ==> expected: <-100.137> but was: <-100.0>
      at app//org.apache.fineract.integrationtests.ClientSavingsIntegrationTest.testRunningBalanceAfterDepositReversalWithBackdateConfigurationOn(ClientSavingsIntegrationTest.java:3029)

File: fineract-provider/src/main/java/org/apache/fineract/portfolio/savings/domain/SavingsAccountTransaction.java [hunk 1]
```diff
@@ -345,6 +348,7 @@ private SavingsAccountTransaction(final SavingsAccount savingsAccount, final Off
         this.reversed = isReversed;
         this.paymentDetail = paymentDetail;
         this.createdDate = createdDate;
+        this.submittedOnDate = DateUtils.getBusinessLocalDate();
```
Comment (adamsaghy) [inline fineract-provider/src/main/java/org/apache/fineract/portfolio/savings/domain/SavingsAccountTransaction.java:351]: There are multiple constructor of the savings transaction. It might worth to cover all places where a transaction can be created then the submitted on date shall be set as well!

**Review by adamsaghy — CHANGES_REQUESTED** [review]
Please see my comment!

File: fineract-provider/src/main/java/org/apache/fineract/portfolio/savings/service/SavingsSchedularInterestPoster.java [hunk 2]
```diff
@@ -206,7 +206,8 @@ private void batchUpdate(final List<SavingsAccountData> savingsAccountDataList)
                             savingsAccountTransactionData.getBalanceNumberOfDays(), savingsAccountTransactionData.getRunningBalance(),
                             savingsAccountTransactionData.getCumulativeBalance(), currentDate, Integer.valueOf(1),
                             savingsAccountTransactionData.isManualTransaction(), savingsAccountTransactionData.getRefNo(),
-                            savingsAccountTransactionData.isReversalTransaction(), savingsAccountTransactionData.getOverdraftAmount(), });
+                            savingsAccountTransactionData.isReversalTransaction(), savingsAccountTransactionData.getOverdraftAmount(),
+                            savingsAccountTransactionData.getTransactionDate() });
```
Comment (adamsaghy) [inline fineract-provider/src/main/java/org/apache/fineract/portfolio/savings/service/SavingsSchedularInterestPoster.java]: this looks wrong to me... submitted on date != transaction date

**Review by adamsaghy — CHANGES_REQUESTED** [review]
Please see my comments!

File: fineract-provider/src/main/java/org/apache/fineract/portfolio/savings/service/SavingsSchedularInterestPoster.java (hunk 2 above)
Comment (Mk9894) [reply fineract-provider/src/main/java/org/apache/fineract/portfolio/savings/service/SavingsSchedularInterestPoster.java]: @adamsaghy I think this should be DateUtils.getBusinessLocalDate() instead of the transaction date. Please correct me if i am wrong

Comment (adamsaghy) [reply fineract-provider/src/main/java/org/apache/fineract/portfolio/savings/service/SavingsSchedularInterestPoster.java]: I believe so

File: fineract-provider/src/main/java/org/apache/fineract/portfolio/savings/domain/SavingsAccountTransaction.java (hunk 1 above)
Comment (marta-jankovics) [reply fineract-provider/src/main/java/org/apache/fineract/portfolio/savings/domain/SavingsAccountTransaction.java:351]: Agreed. It is a wrong pattern having multiple constructors in the middle of the class, because it leads to these mistakes. Better to have only one at the beginning of the class (or somewhere agreed on generally)

File: fineract-provider/src/main/java/org/apache/fineract/portfolio/savings/service/SavingsSchedularInterestPoster.java [hunk 3]
```diff
@@ -206,7 +206,8 @@ private void batchUpdate(final List<SavingsAccountData> savingsAccountDataList)
                             savingsAccountTransactionData.getBalanceNumberOfDays(), savingsAccountTransactionData.getRunningBalance(),
```
Comment (marta-jankovics) [inline fineract-provider/src/main/java/org/apache/fineract/portfolio/savings/service/SavingsSchedularInterestPoster.java:206]: This change is ok, but checked a bit the usages of SavingsAccountTransactionData and found several issues. Started to collect but there are more.
Please rename SavingsAccountTransactionData constructor parameter createdDate
SavingsAccountTransactionData line 522: this.submittedOnDate = transactionDate; 
does not seem to be correct
SavingsAccountReadPlatformServiceImpl.extractData line 603:
final LocalDate transSubmittedOnDate = JdbcSupport.getLocalDate(rs, "createdDate");
not correct
RecurringAccountDepositTransactionTemplateMapper sends m_mandatory_savings_schedule.duedate as submittedOnDate - is this correct?
SavingsAccountTransactionsMapper in DepositAccountReadPlatformServiceImpl sends transaction_date - not correct
and this init is used from 8 other places
Please check all the usages of SavingsAccountTransactionData static initialisers (all of them), even if it is  it is hard to follow.

Other place to check for submittedOnDate:
SavingsAccountReadPlatformServiceImpl.SavingsAccountTransactionsMapper
            sqlBuilder.append("tr.created_date as submittedOnDate,");

File: fineract-provider/src/main/java/org/apache/fineract/portfolio/savings/domain/SavingsAccountTransaction.java [hunk 4]
```diff
@@ -365,6 +369,7 @@ private SavingsAccountTransaction(final Integer transactionTypeEnum, final Local
         this.balanceNumberOfDays = balanceNumberOfDays;
         this.overdraftAmount = overdraftAmount;
         this.createdDate = createdDate;
+        this.submittedOnDate = DateUtils.getBusinessLocalDate();
```
Comment (adamsaghy) [inline fineract-provider/src/main/java/org/apache/fineract/portfolio/savings/domain/SavingsAccountTransaction.java:372]: Please check the copy transaction logic (`copyTransaction` method) as well, whether it gets properly submitted on date or not!

**Review by adamsaghy — DISMISSED** [review]
LGTM

**Review by marta-jankovics — CHANGES_REQUESTED** [review]

**Review by marta-jankovics — CHANGES_REQUESTED** [review]
Please see my comments. even if I request change it does not change the status of the PR, but please consider it not approved.

**Review by adamsaghy — CHANGES_REQUESTED** [review]
Please see Marta's comments!

File: fineract-provider/src/main/java/org/apache/fineract/portfolio/savings/service/SavingsSchedularInterestPoster.java (hunk 3 above)
Comment (Mk9894) [reply fineract-provider/src/main/java/org/apache/fineract/portfolio/savings/service/SavingsSchedularInterestPoster.java:206]: @marta-jankovics I planned to do this part in the next PR. This PR just introduces submittedOnDate and initializes it. This PR might become huge if we make changes to all places where submittedOnDate is consumed.

What do you think?

**Review by marta-jankovics — APPROVED** [review]
We can merge this PR in this state and have the requested changes in a separate PR, because this PR is blocking other PRs to be merged.
